### PR TITLE
Added click between steps capability

### PR DESCRIPTION
**Pull request by wilsonjs**
Already visited steps are clickable.

Refactored Stepper from pure function component to component class to access lifecycle methods, added unit tests for nav between visited steps.

### CODE REVIEW

File: src/stepper/Stepper.js [hunk 1]
```diff
@@ -8,15 +9,54 @@ function clamp(from, to, value) {
...
 /* eslint-disable react/no-array-index-key */
-const Stepper = ({ steps, activeStep }) => {
-  const activeStepIndex = clamp(0, steps.length - 1, activeStep);
-  const stepPercentage = 1 / (steps.length - 1);
-  const percentageCompleted = activeStepIndex / (steps.length - 1);
-  const filledWidth = Math.max(percentageCompleted - stepPercentage, 0);
-  const endingWidth = Math.min(activeStepIndex, 1) * stepPercentage;
-
-  const renderStep = (step, index) => {
+class Stepper extends Component {
```
Comment (wilsonjs) [inline src/stepper/Stepper.js]: Changed to class in order to access lifecycle methods and be able to update `highestVisitedIndex` value on the state.

File: src/stepper/Stepper.js [hunk 2]
```diff
@@ -8,15 +9,53 @@ function clamp(from, to, value) {
 }
 
 /* eslint-disable react/no-array-index-key */
-const Stepper = ({ steps, activeStep }) => {
-  const activeStepIndex = clamp(0, steps.length - 1, activeStep);
-  const stepPercentage = 1 / (steps.length - 1);
-  const percentageCompleted = activeStepIndex / (steps.length - 1);
-  const filledWidth = Math.max(percentageCompleted - stepPercentage, 0);
-  const endingWidth = Math.min(activeStepIndex, 1) * stepPercentage;
+class Stepper extends Component {
```
Comment (wilsonjs) [inline src/stepper/Stepper.js]: Changed from pure component to class to access lifecycle methods and update state value: `highestVisitedIndex`,

File: src/stepper/Stepper.js [hunk 3]
```diff
@@ -27,22 +66,17 @@ const Stepper = ({ steps, activeStep }) => {
     const labelButton = (
       <button
         className="btn-unstyled tw-stepper__step-label"
-        disabled={!clickable}
-        onClick={() => clickable && step.onClick()}
+        disabled={disabled}
```
Comment (wilsonjs) [inline src/stepper/Stepper.js]: Button will be enabled when not-active, clickable or visited.

File: src/stepper/Stepper.js [hunk 4]
```diff
@@ -8,15 +9,53 @@ function clamp(from, to, value) {
...
+    this.updateState(nextProps);
+  }
+
+  updateState({ steps, activeStep }) {
+    const activeStepIndex = clamp(0, steps.length - 1, activeStep);
+    const stepPercentage = 1 / (steps.length - 1);
+    const percentageCompleted = activeStepIndex / (steps.length - 1);
+    const filledWidth = Math.max(percentageCompleted - stepPercentage, 0);
+    const endingWidth = Math.min(activeStepIndex, 1) * stepPercentage;
+    const highestVisitedIndex = Math.max(this.state.highestVisitedIndex, activeStepIndex);
```
Comment (wilsonjs) [inline src/stepper/Stepper.js]: All these constants are the same as before, only the `highestVisitedIndex` is new.

File: src/stepper/Stepper.js [hunk 5]
```diff
@@ -27,22 +66,17 @@ const Stepper = ({ steps, activeStep }) => {
...
       <li
         key={index}
-        style={{ left: `${index * stepPercentage * 100}%` }}
-        className={`
-          hidden-xs
-          tw-stepper__step
-          ${index === activeStepIndex ? 'tw-stepper__step--active' : ''}
-          ${clickable ? 'tw-stepper__step--clickable' : ''}
-        `}
+        style={{ left: `${index * this.state.stepPercentage * 100}%` }}
```
Comment (mstaicu) [inline src/stepper/Stepper.js]: I would precompute all the values so that I have as little business logic in my markup as possible.

**Comment by Tankenstein** [comment]
Please don’t merge until i review, at a layover right now, will review in the evening/morning :)

File: src/stepper/Stepper.js (hunk 5 above)
Comment (wilsonjs) [reply src/stepper/Stepper.js]: Fair point, was just trying to modify the code as little as possible.

**Pull request closed without merging**